File: benchmarks/agent/analysis/records.go
// Package analysis turns an agent benchmark results tree into a report in
// three stages: Extract (runs to metrics), Analyze (metrics to statistics)
// and Render (statistics to markdown). Each stage only reads artifacts, so a
// scored run is analyzed as often as wanted without re-running an agent. The
// record types here are the file formats: metrics.jsonl carries one ScoredRun
// or ControlRun per line, analysis.json one Analysis.
package analysis

import "github.com/egladman/magus/benchmarks/agent/analysis/pycompat"

// Number is a JSON number that remembers whether it was an int or a float.
type Number = pycompat.Number

// The two recipes the paired comparison names. A record's arm stays a plain
// string: the runner accepts arms outside this pair (_selftest), and cells
// and pass rates cover whatever ran.
const (
	ArmFull    = "full"
	ArmRampant = "rampant"
)

// Control kinds a meta.json may carry.
const (
	ControlGolden = "golden"
	ControlNull   = "null"
)

// Verdicts a paired delta can earn.
const (
	VerdictLower        = "lower under full"
	VerdictHigher       = "higher under full"
	VerdictInconclusive = "inconclusive"
)

// TokenCounts are the billed tokens of one run. TotalBilled is a token count,
// not a price.
type TokenCounts struct {
	Input       int64 `json:"input"`
	Output      int64 `json:"output"`
	CacheRead   int64 `json:"cache_read"`
	CacheWrite  int64 `json:"cache_write"`
	TotalBilled int64 `json:"total_billed"`
}

// GuardEvents is guard activity read from the magus trail. A run carries nil
// instead when no trail was captured; zero means the guard was present and
// silent, and the two must never be conflated.
type GuardEvents struct {
	Denials    int64 `json:"denials"`
	Advisories int64 `json:"advisories"`
	SkillLoads int64 `json:"skill_loads"`
}

// InvariantViolations lists what a run's diff did that no task allows.
type InvariantViolations struct {
	TestsDeleted []string `json:"tests_deleted"`
}

// ControlRun is a golden or null control: identity, kind and check verdict,
// no agent.
type ControlRun struct {
	RunID      string  `json:"run_id"`
	Arm        string  `json:"arm"`
	Task       string  `json:"task"`
	Rep        int64   `json:"rep"`
	Model      string  `json:"model"`
	Control    string  `json:"control"`
	ExitReason *string `json:"exit_reason"`
	CheckExit  *int64  `json:"check_exit"`
	Success    *bool   `json:"success"`
}

// ScoredRun is one agent run measured from its artifacts. Success is nil when
// no check ran and GuardEvents is nil when no trail was captured; neither is a
// zero. Dollars is the host's billed cost when the transcript recorded one and
// the priced table total otherwise. Control is always nil: every row carries
// the key so a reader tells the two record kinds apart by it.
type ScoredRun struct {
	RunID                string              `json:"run_id"`
	Arm                  string              `json:"arm"`
	Task                 string              `json:"task"`
	Rep                  int64               `json:"rep"`
	Model                string              `json:"model"`
	Tokens               TokenCounts         `json:"tokens"`
	Dollars              float64             `json:"dollars"`
	TableDollarsUSD      float64             `json:"table_dollars_usd"`
	ReportedCostUSD      *float64            `json:"reported_cost_usd"`
	CacheWriteTTLAssumed bool                `json:"cache_write_ttl_assumed"`
	Turns                int64               `json:"turns"`
	ToolCalls            int64               `json:"tool_calls"`
	ToolCallsByName      map[string]int64    `json:"tool_calls_by_name"`
	FileReads            int64               `json:"file_reads"`
	DistinctFilesRead    int64               `json:"distinct_files_read"`
	ReReadRate           float64             `json:"re_read_rate"`
	ToolResultBytes      int64               `json:"tool_result_bytes"`
	GuardEvents          *GuardEvents        `json:"guard_events"`
	CheckExit            *int64              `json:"check_exit"`
	Success              *bool               `json:"success"`
	InvariantViolations  InvariantViolations `json:"invariant_violations"`
	WallMs               *Number             `json:"wall_ms"`
	TimeToFirstEditMs    *Number             `json:"time_to_first_edit_ms"`
	TimeToDoneMs         *Number             `json:"time_to_done_ms"`
	Effort               *string             `json:"effort"`
	MaxTurns             *Number             `json:"max_turns"`
	BudgetUSD            *Number             `json:"budget_usd"`
	MagusBinary          *string             `json:"magus_binary"`
	MagusVersion         *string             `json:"magus_version"`
	FixtureSha           *string             `json:"fixture_sha"`
	Started              *string             `json:"started"`
	Ended                *string             `json:"ended"`
	ExitReason           *string             `json:"exit_reason"`
	Control              *string             `json:"control"`
}

// RunRecord is one line of metrics.jsonl: exactly one of the two is set.
type RunRecord struct {
	Scored  *ScoredRun
	Control *ControlRun
}

// Spread is median and IQR beside the mean; a mean alone hides the spread
// that matters. Median and the IQR bounds keep the int form of a count metric
// when they fall on a value rather than between two.
type Spread struct {
	N      int64      `json:"n"`
	Median *Number    `json:"median"`
	IQR    [2]*Number `json:"iqr"`
	Mean   *float64   `json:"mean"`
}

// CellStats are the pass rates and metric spreads of one (arm, task) cell.
type CellStats struct {
	N                  int64             `json:"n"`
	Successes          int64             `json:"successes"`
	UnknownOutcomes    int64             `json:"unknown_outcomes"`
	PassAt1            *float64          `json:"pass_at_1"`
	PassAt1CI          [2]*float64       `json:"pass_at_1_ci"`
	K                  int64             `json:"k"`
	PassPowK           float64           `json:"pass_pow_k"`
	Metrics            map[string]Spread `json:"metrics"`
	MetricsSuccessOnly map[string]Spread `json:"metrics_success_only"`
}

// CostOfPass is expected dollars per correct solution; infinite at a zero
// pass rate.
type CostOfPass struct {
	Value    *float64 `json:"value"`
	Infinite bool     `json:"infinite"`
}

// ArmSummary is one arm's pass rate and cost-of-pass over every task.
type ArmSummary struct {
	N             int64       `json:"n"`
	Successes     int64       `json:"successes"`
	PassRate      float64     `json:"pass_rate"`
	PassRateCI    [2]*float64 `json:"pass_rate_ci"`
	Dollars       Spread      `json:"dollars"`
	CostOfPassUSD CostOfPass  `json:"cost_of_pass_usd"`
}

// PairedDelta is treatment minus baseline for one metric on one task, over
// matched reps.
type PairedDelta struct {
	NPairs         int64    `json:"n_pairs"`
	DeltaMean      *float64 `json:"delta_mean"`
	DeltaMedian    *Number  `json:"delta_median"`
	CILow          *float64 `json:"ci_low"`
	CIHigh         *float64 `json:"ci_high"`
	P              *float64 `json:"p"`
	BaselineMedian *Number  `json:"baseline_median"`
	Relative       *float64 `json:"relative"`
	CIExcludesZero bool     `json:"ci_excludes_zero"`
	Verdict        string   `json:"verdict"`
	PHolm          *float64 `json:"p_holm"`
}

// ControlCount is how many controls of one kind a task ran and how many passed.
type ControlCount struct {
	N      int64    `json:"n"`
	Passes int64    `json:"passes"`
	RunIDs []string `json:"run_ids"`
}

// ControlCell says whether one task's check discriminates: golden must pass,
// null must fail.
type ControlCell struct {
	Golden        ControlCount `json:"golden"`
	Null          ControlCount `json:"null"`
	Discriminates bool         `json:"discriminates"`
}

// UnpairedRep is a rep one arm ran and another did not.
type UnpairedRep struct {
	Task        string   `json:"task"`
	Rep         int64    `json:"rep"`
	MissingArms []string `json:"missing_arms"`
}

// DataQuality holds the facts the report's caveats are generated from, not
// prose about them.
type DataQuality struct {
	TableToBilledRatioMedian    *float64      `json:"table_to_billed_ratio_median"`
	RunsWithoutBilledCost       []string      `json:"runs_without_billed_cost"`
	RunsWithoutGuardEvents      []string      `json:"runs_without_guard_events"`
	RunsWithoutCheck            []string      `json:"runs_without_check"`
	RunsWithAssumedCacheTTL     []string      `json:"runs_with_assumed_cache_ttl"`
	RunsWithInvariantViolations []string      `json:"runs_with_invariant_violations"`
	IncompletePairs             []UnpairedRep `json:"incomplete_pairs"`
}

// Analysis is analysis.json. Cells are keyed arm/task; Paired is keyed
// metric, then task.
type Analysis struct {
	Seed             int64                             `json:"seed"`
	BootstrapIters   int64                             `json:"bootstrap_iters"`
	MinRelativeDelta float64                           `json:"min_relative_delta"`
	Runs             int64                             `json:"runs"`
	Arms             []string                          `json:"arms"`
	Tasks            []string                          `json:"tasks"`
	Models           []string                          `json:"models"`
	Controls         map[string]ControlCell            `json:"controls"`
	Cells            map[string]CellStats              `json:"cells"`
	ArmSummary       map[string]ArmSummary             `json:"arm_summary"`
	Paired           map[string]map[string]PairedDelta `json:"paired"`
	DataQuality      DataQuality                       `json:"data_quality"`
}

File: benchmarks/agent/analysis/records.py
"""Data shapes the analysis stages hand each other, and their JSON boundary.

extract.py writes one run record per line of metrics.jsonl; analyze.py reads
those back and writes one Analysis to analysis.json; report.py reads that.
Every shape is a frozen dataclass, and `to_json` / `from_json` are the only
places a stage meets a dict, so the file formats are these classes' fields.
"""

from __future__ import annotations

import dataclasses
import enum
import types
import typing
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Self


class Arm(enum.StrEnum):
    """The two recipes the paired comparison names.

    A record's arm stays a plain string: the runner accepts arms outside this
    pair (`_selftest`), and cells and pass rates cover whatever ran.
    """

    FULL = "full"
    RAMPANT = "rampant"


class ControlKind(enum.StrEnum):
    GOLDEN = "golden"
    NULL = "null"


class Verdict(enum.StrEnum):
    LOWER = "lower under full"
    HIGHER = "higher under full"
    INCONCLUSIVE = "inconclusive"


class JsonRecord:
    """Mixin for the frozen dataclasses below: nested fields out, nested fields in.

    `from_json` rebuilds a value from its field annotation: a JsonRecord, an
    enum, `X | None`, `tuple[X, ...]` and `dict[str, X]` are decoded, any other
    annotation is taken as the JSON scalar it already is. A field with a default
    may be absent from the data; any other missing field raises KeyError.
    """

    def to_json(self) -> dict[str, Any]:
        return dataclasses.asdict(self)  # type: ignore[call-overload]

    @classmethod
    def from_json(cls, data: Mapping[str, Any]) -> Self:
        hints = typing.get_type_hints(cls)
        values: dict[str, Any] = {}
        for field in dataclasses.fields(cls):  # type: ignore[arg-type]
            if field.name in data:
                values[field.name] = _decode(hints[field.name], data[field.name])
            elif _required(field):
                raise KeyError(field.name)
        return cls(**values)


def _required(field: dataclasses.Field[Any]) -> bool:
    return field.default is dataclasses.MISSING and field.default_factory is dataclasses.MISSING


def _decode(hint: Any, value: Any) -> Any:
    origin = typing.get_origin(hint)
    if origin in (types.UnionType, typing.Union):
        if value is None:
            return None
        inner = next(arg for arg in typing.get_args(hint) if arg is not type(None))
        return _decode(inner, value)
    if origin is tuple:
        return tuple(_decode(typing.get_args(hint)[0], item) for item in value)
    if origin is dict:
        return {key: _decode(typing.get_args(hint)[1], item) for key, item in value.items()}
    if isinstance(hint, type) and issubclass(hint, JsonRecord):
        return hint.from_json(value)
    if isinstance(hint, type) and issubclass(hint, enum.Enum):
        return hint(value)
    return value


@dataclass(frozen=True)
class TokenCounts(JsonRecord):
    """Billed tokens of one run. total_billed is a token count, not a price."""

    input: int
    output: int
    cache_read: int
    cache_write: int
    total_billed: int


@dataclass(frozen=True)
class GuardEvents(JsonRecord):
    """Guard activity read from the magus trail.

    A run carries None instead when no trail was captured; zero means the guard
    was present and silent, and the two must never be conflated.
    """

    denials: int
    advisories: int
    skill_loads: int


@dataclass(frozen=True)
class InvariantViolations(JsonRecord):
    tests_deleted: tuple[str, ...] = ()


@dataclass(frozen=True)
class ControlRun(JsonRecord):
    """A golden or null control: identity, kind and check verdict, no agent."""

    run_id: str
    arm: str
    task: str
    rep: int
    model: str
    control: ControlKind
    exit_reason: str | None
    check_exit: int | None
    success: bool | None


@dataclass(frozen=True)
class ScoredRun(JsonRecord):
    """One agent run measured from its artifacts.

    `success` is None when no check ran, and `guard_events` is None when no
    trail was captured; neither is a zero. `dollars` is the host's billed cost
    when the transcript recorded one and the priced table total otherwise.
    """

    run_id: str
    arm: str
    task: str
    rep: int
    model: str
    tokens: TokenCounts
    dollars: float
    table_dollars_usd: float
    reported_cost_usd: float | None
    cache_write_ttl_assumed: bool
    turns: int
    tool_calls: int
    tool_calls_by_name: dict[str, int]
    file_reads: int
    distinct_files_read: int
    re_read_rate: float
    tool_result_bytes: int
    guard_events: GuardEvents | None
    check_exit: int | None
    success: bool | None
    invariant_violations: InvariantViolations
    wall_ms: int | None
    time_to_first_edit_ms: int | None
    time_to_done_ms: int | None
    effort: str | None = None
    max_turns: int | None = None
    budget_usd: float | None = None
    magus_binary: str | None = None
    magus_version: str | None = None
    fixture_sha: str | None = None
    started: str | None = None
    ended: str | None = None
    exit_reason: str | None = None

    def to_json(self) -> dict[str, Any]:
        # Every row carries the key so a reader tells the two kinds apart by it.
        return {**super().to_json(), "control": None}


RunRecord = ScoredRun | ControlRun


def run_from_json(data: Mapping[str, Any]) -> RunRecord:
    if data.get("control"):
        return ControlRun.from_json(data)
    return ScoredRun.from_json(data)


@dataclass(frozen=True)
class Spread(JsonRecord):
    """Median and IQR beside the mean; a mean alone hides the spread that matters."""

    n: int
    median: float | None
    iqr: tuple[float | None, float | None]
    mean: float | None


@dataclass(frozen=True)
class CellStats(JsonRecord):
    """Pass rates and metric spreads of one (arm, task) cell."""

    n: int
    successes: int
    unknown_outcomes: int
    pass_at_1: float | None
    pass_at_1_ci: tuple[float | None, float | None]
    k: int
    pass_pow_k: float
    metrics: dict[str, Spread]
    metrics_success_only: dict[str, Spread]


@dataclass(frozen=True)
class CostOfPass(JsonRecord):
    """Expected dollars per correct solution; infinite at a zero pass rate."""

    value: float | None
    infinite: bool


@dataclass(frozen=True)
class ArmSummary(JsonRecord):
    n: int
    successes: int
    pass_rate: float
    pass_rate_ci: tuple[float | None, float | None]
    dollars: Spread
    cost_of_pass_usd: CostOfPass


@dataclass(frozen=True)
class PairedDelta(JsonRecord):
    """Treatment minus baseline for one metric on one task, over matched reps."""

    n_pairs: int
    delta_mean: float | None
    delta_median: float | None
    ci_low: float | None
    ci_high: float | None
    p: float | None
    baseline_median: float | None
    relative: float | None
    ci_excludes_zero: bool
    verdict: Verdict
    p_holm: float | None = None


@dataclass(frozen=True)
class ControlCount(JsonRecord):
    n: int
    passes: int
    run_ids: tuple[str, ...]


@dataclass(frozen=True)
class ControlCell(JsonRecord):
    """Whether one task's check discriminates: golden must pass, null must fail."""

    golden: ControlCount
    null: ControlCount
    discriminates: bool


@dataclass(frozen=True)
class UnpairedRep(JsonRecord):
    task: str
    rep: int
    missing_arms: tuple[str, ...]


@dataclass(frozen=True)
class DataQuality(JsonRecord):
    """Facts the report's caveats are generated from, not prose about them."""

    table_to_billed_ratio_median: float | None
    runs_without_billed_cost: tuple[str, ...]
    runs_without_guard_events: tuple[str, ...]
    runs_without_check: tuple[str, ...]
    runs_with_assumed_cache_ttl: tuple[str, ...]
    runs_with_invariant_violations: tuple[str, ...]
    incomplete_pairs: tuple[UnpairedRep, ...]


@dataclass(frozen=True)
class Analysis(JsonRecord):
    """analysis.json. Cells are keyed `arm/task`; paired is keyed metric, then task."""

    seed: int
    bootstrap_iters: int
    min_relative_delta: float
    runs: int
    arms: tuple[str, ...]
    tasks: tuple[str, ...]
    models: tuple[str, ...]
    controls: dict[str, ControlCell]
    cells: dict[str, CellStats]
    arm_summary: dict[str, ArmSummary]
    paired: dict[str, dict[str, PairedDelta]]
    data_quality: DataQuality
